### PR TITLE
feat: [CDS-45754]:Supporting rbac based paginated filtering for Service/Environment/Template list API 

### PR DESCRIPTION
**Pull request by vivekkarnatiharness**
## Describe your changes
Have made changes to show only the entities that user is having is access to. 
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java [hunk 1]
```diff
@@ -520,7 +523,42 @@ private ResponseDTO<PageResponse<EnvironmentResponse>> getEnvironmentsPageByCrit
     } else {
       pageRequest = PageUtils.getPageRequest(page, size, sort);
     }
-    return getEnvironmentsPageByCriteria(criteria, pageRequest);
+
+    Page<Environment> environmentPage = null;
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java:527]: this can be final

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java [hunk 2]
```diff
@@ -520,7 +523,42 @@ private ResponseDTO<PageResponse<EnvironmentResponse>> getEnvironmentsPageByCrit
...
+      }
+
+      populateInFilter(criteria, EnvironmentKeys.identifier,
+          environmentList.stream().map(Environment::getIdentifier).collect(toList()));
+
+      environmentPage = environmentService.list(criteria, pageRequest);
+    }
+
+    environmentPage.forEach(environment -> {
+      if (EmptyPredicate.isEmpty(environment.getYaml())) {
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java:554]: you can use the method environment.fetchNonEmptyYaml(.) for this

File: 120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java (hunk 1 above)
Comment (vivekkarnatiharness) [reply 120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java:525]: If we declare this final,we can't assign another value right?

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 36 at r5](https://reviewable.io/reviews/harness/harness-core/45652#-NRg8S2448Rll6die_AR:-NRg8S2448Rll6die_AS:b5ldqdd) ([raw file](https://github.com/harness/harness-core/blob/dc34619c3432c639157252f4bae1f6b1c5caa05b/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L36)):*
> ```Java
>   @Inject private AccessControlClient accessControlClient;
>   private boolean hasProdAccess = false;
>   private boolean hasPreProdAccess = false;
> ```

singleton classes MUST not have any value fields. Same instance of the class is going to used by multiple threads

**Review by vivekkarnatiharness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)* status: 8 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 36 at r5](https://reviewable.io/reviews/harness/harness-core/45652#-NRg8S2448Rll6die_AR:-NRgvXIiA-ZbyXQ9VWdg:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/dc34619c3432c639157252f4bae1f6b1c5caa05b/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L36)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

singleton classes MUST not have any value fields. Same instance of the class is going to used by multiple threads
</blockquote></details>

Done.

**Review by vivekkarnatiharness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)* status: 8 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 527 at r1](https://reviewable.io/reviews/harness/harness-core/45652#-NRW7ZyqSA-0ySm0alLX-r1-527:-NRgvZIt8FFGNjnIuyb4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9ec9341b15180671267dbe0a22e2c140f88d3f2e/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L527)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Correct
Final Page<Environment> environmentPage;

if(..) {
 page = ..
} else {
 page = ...
}
</blockquote></details>

Done.

**Review by vivekkarnatiharness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)* status: 8 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 529 at r1](https://reviewable.io/reviews/harness/harness-core/45652#-NRW7ZyqSA-0ySm0alLX-r1-529:-NRgv_U20KM_GoAs_NVr:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9ec9341b15180671267dbe0a22e2c140f88d3f2e/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L529)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

better name would be hasViewPermissionForAllEnvironments();
</blockquote></details>

Done.

**Review by vivekkarnatiharness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)* status: 8 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 541 at r1](https://reviewable.io/reviews/harness/harness-core/45652#-NRW7ZyqSA-0ySm0alLX-r1-541:-NRgvbF83cVWQIOGO4kU:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9ec9341b15180671267dbe0a22e2c140f88d3f2e/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L541)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we directly not do this ? instead of creating a local variable 

List<Environment> environmentList = environmentRbacHelper.getPermittedEnvironmentsList(environmentPages.getContent())
</blockquote></details>

Done.

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 71 at r5](https://reviewable.io/reviews/harness/harness-core/45652#-NRijHDID7T1-z6iViFT:-NRijHDID7T1-z6iViFU:b-kspiw2) ([raw file](https://github.com/harness/harness-core/blob/dc34619c3432c639157252f4bae1f6b1c5caa05b/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L71)):*
> ```Java
>                                  environments.get(0).getOrgIdentifier(), environments.get(0).getProjectIdentifier()))
>                              .resourceType(NGResourceType.ENVIRONMENT)
>                              .build());
> ```

This does not look correct. Users can be assigned permissions to environment types based on scopes. for eg. production environments in the scope project but only pre prod environments at the account level 

I understand that this works only when listing environments from same scope, so lets document this as a java doc in this method

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 554 at r1](https://reviewable.io/reviews/harness/harness-core/45652#-NRW7ZyqSA-0ySm0alLX-r1-554:-NRmLew201QE1XcqnMY3:ba9jcf1) ([raw file](https://github.com/harness/harness-core/blob/9ec9341b15180671267dbe0a22e2c140f88d3f2e/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L554)):*
<details><summary><i>Previously, vivekkarnatiharness wrote…</i></summary><blockquote>

Done.
</blockquote></details>

we are not using the return value for this method ?

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 48 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmLlX1A-yBJKTfPtGh:-NRmLlX1A-yBJKTfPtGi:b3u2fyk) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L48)):*
> ```Java
> 
>     final boolean hasProdAccess;
>     final boolean hasPreProdAccess;
> ```

do we need to define these 2 variables here ? can you check the PR I shared ?

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)* status: 9 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 69 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmMQFFEZIRPxgDAb1U:-NRmMQFFEZIRPxgDAb1V:b-f3tso7) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L69)):*
> ```Java
> 
>     // pre-prod permission check
> ```

lets get rid of these new lines
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 79 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmMTbMBt43Sh2HEbm0:-NRmMTbMBt43Sh2HEbm1:b-f3tso7) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L79)):*
> ```Java
> 
>     // Prod permission check
> ```

lets get rid of these new lines
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 124 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmMq2qAO5I4K24Mocr:-NRmMq2rBEIPPXxQWLXG:b-w0cedj) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L124)):*
> ```Java
>   private List<AccessControlDTO> removeTypeBasedAccessControlDTOs(List<AccessControlDTO> accessControlDTOList) {
>     return accessControlDTOList.stream()
>         .filter(dto -> isEmpty(dto.getResourceAttributes()) || isEmpty(dto.getResourceAttributes().get("type")))
> ```

We can extract this out as a constant



_Code quote:_
```Java
"type"
```
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 128 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmN-XC5lbRUp-1S7pq:-NRmN-XC5lbRUp-1S7pr:b5qs8ei) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L128)):*
> ```Java
>   }
> 
>   private EnvironmentTypeFilteredResponse CheckingTypeBasedFilters(List<AccessControlDTO> accessControlDTOList) {
> ```

method name should start with small letters

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)* status: 9 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 1117 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmNVrl15GPU2RIfEXU:-NRmNVrl15GPU2RIfEXV:b-d0wxhj) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L1117)):*
> ```Java
>       throw new InvalidRequestException(ex.getMessage());
>     }
>   }
> ```

new line

**Review by vivekkarnatiharness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45652)* status: 14 file(s) to review. 1 change requests. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 48 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmLlX1A-yBJKTfPtGh:-NRpuA-UEF-IG2_ykw0p:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L48)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

do we need to define these 2 variables here ? can you check the PR I shared ?
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 69 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmMQFFEZIRPxgDAb1U:-NRpuAsWEYBzDEtcgytn:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L69)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets get rid of these new lines
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 79 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmMTbMBt43Sh2HEbm0:-NRpuBYE5Tp5W-qwhVc6:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L79)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets get rid of these new lines
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 124 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmMq2qAO5I4K24Mocr:-NRpuDJFEmYgtie2muTT:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L124)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

We can extract this out as a constant
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 128 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmN-XC5lbRUp-1S7pq:-NRpuFeRES84NhDL3_q2:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L128)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

method name should start with small letters
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 554 at r1](https://reviewable.io/reviews/harness/harness-core/45652#-NRW7ZyqSA-0ySm0alLX-r1-554:-NRpuGDF2A57zLPMuluN:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9ec9341b15180671267dbe0a22e2c140f88d3f2e/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L554)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

we are not using the return value for this method ?
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 543 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmNJem4V9pTv-CarYX:-NRpuHQz-WazPRwKEiPV:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L543)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

we should set it ? 
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 1117 at r7](https://reviewable.io/reviews/harness/harness-core/45652#-NRmNVrl15GPU2RIfEXU:-NRpuJEc2gRThln09i5v:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5aa0998f928ba753f34e67827ce1bb238d87c4ab/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L1117)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

new line
</blockquote></details>

Done.

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java` line 159 at r11](https://reviewable.io/reviews/harness/harness-core/45652#-NRq5DmK0bBFHWAf2Jx1:-NRq5DmK0bBFHWAf2Jx2:b-3j3qja) ([raw file](https://github.com/harness/harness-core/blob/7533d4beabbfc5c34e29083b75374dd3362d7300/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentRbacHelper.java#L159)):*
> ```Java
>   private Map<String, String> getEnvironmentAttributesMap(String environmentType) {
>     Map<String, String> environmentAttributes = new HashMap<>();
>     environmentAttributes.put("type", environmentType);
> ```

please extract out this constant as well as done at other places in this class

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentTypeFilteredResponse.java` line 15 at r11](https://reviewable.io/reviews/harness/harness-core/45652#-NRq5OcECsIHLo8ZnGZN:-NRq5OcECsIHLo8ZnGZO:bpp2uqn) ([raw file](https://github.com/harness/harness-core/blob/7533d4beabbfc5c34e29083b75374dd3362d7300/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentTypeFilteredResponse.java#L15)):*
> ```Java
> import lombok.Builder;
> import lombok.Data;
> import lombok.NoArgsConstructor;
> ```

nit: new line here for readability

**Pull request closed without merging**